src/magic_folder/test/test_local_snapshot.py

```diff
@@ -0,0 +1,336 @@
+            file = self.magic_path.child(filename)
+            with file.open("wb") as f:
+                f.write(content)
+            files.append(file)
+
+        self.snapshot_service.startService()
+
+        list_d = []
+        for file in files:
+            result_d = self.snapshot_service.add_file(file)
+            list_d.append(result_d)
+
+        d = defer.gatherResults(list_d)
+
+        self.assertThat(
+            d,
+            succeeded(Always()),
+        )
+
+        self.assertThat(
+            self.snapshot_service.stopService(),
+            succeeded(Always())
+        )
+
+        self.assertThat(
+            len(self._snapshot_creator.processed.keys()),
+            Equals(len(files))
+        )
+
+        self.assertThat(
+            sorted(self._snapshot_creator.processed.keys()),
+            Equals(sorted(files))
+        )
+
+    @given(content=binary())
+    def test_add_file_failures(self, content):
+        """
+        Test with bad inputs to check failure paths.
+        """
+        foo = self.magic_path.child("foo")
+        with foo.open("wb") as f:
+            f.write(content)
+
+        self.snapshot_service.startService()
+
+        # try adding a string that represents the path
+        with ExpectedException(TypeError,
+                               "argument must be a FilePath"):
+            self.snapshot_service.add_file(foo.path)
+
+        # try adding a directory
+        tmpdir = FilePath(self.mktemp())
+        bar_dir = self.magic_path.child(tmpdir.basename())
+        bar_dir.makedirs()
+
+        with ExpectedException(ValueError,
+                               "expected a file"):
+            self.snapshot_service.add_file(bar_dir)
+
+
+        # try adding a file outside the magic folder directory
+        tmpfile = FilePath(self.mktemp())
+        with tmpfile.open("wb") as f:
+            f.write(content)
+
+        with ExpectedException(ValueError,
+                               "The path being added .*"):
+            self.snapshot_service.add_file(tmpfile)
+
+        self.assertThat(
+            self.snapshot_service.stopService(),
+            succeeded(Always())
+        )
+
+    @given(content1=binary(min_size=1),
+           content2=binary(min_size=1),
+           filename=path_segments().map(lambda p: p.encode("utf-8")),
+    )
+    def test_add_a_file_twice(self, filename, content1, content2):
+        """
+        Adding a file twice should create two local snapshots with the first
+        one being the parent of the second.
+        """
+        foo = self.magic_path.child(filename)
+        with foo.open("wb") as f:
+            f.write(content1)
+
+        self.snapshot_service.startService()
+        self.snapshot_service.add_file(foo)
+
+        with foo.open("wb") as f:
+            f.write(content2)
+
+        # it should use the previous localsnapshot as its parent.
+        self.snapshot_service.add_file(foo)
```

Comment:
I think there should be an assertion that the previous add_file call succeeded somewhere before this assertion.  If the first add_file hasn't happened yet then there will be no parent for the second.